### ISSUE TITLE
TypeQL fetch queries

### ISSUE THREAD

**Issue by flyingsilverfin** [vaticle/typedb#6857]
## Problem to Solve
We introduce a `TypeQLFetch` query, which returns a native JSON data structure and allows creating and retrieving nested data generated by expanding the concepts in a 'match' query.

## Current Workaround
Use the Concept API or multiple round trips to fetch multiple layers of data or attached attributes. Note `fetch` will only return concrete _values_ built from concepts, but not TypeDB's typical concepts. In other words, every entity/relation will need to be converted to attributes to be meaningfully returned since IIDs will not be provided.

## Proposed Solution
Introduce 'fetch' as an alternative to 'get' for 3 useage patterns:
1) mapping concepts as attributes:

```
match
$x isa person, has email "abc@gmail.com";
$y isa company, has company-id 1234;
$r ($x, $y) isa employment;
fetch
$x as person: name, age, address;  # fetch names, ages, and addresses for a the person
$y as company: address;            # fetch addresses for the company
$r as employment: start-date;      # fetch start-dates for the company
```

This should return JSON:

```
[ 
{
  person: {
    type: 'person',   // guaranteed key
    name: [ {type: 'name', value: 'john johnson'}, {type: 'nickname', value: 'johnny'}],
    age: [{type: 'age', value: 55}],
    address: [{type: 'address', value: '123 abc street'}]
  },
  company: {
    type: 'nonprofit' // respects polymorphism
    address: [ {type: 'address', value: '1 root street'}, { type: 'address', value: '2 central value'}, ...]
  },
  employment: {
    type: 'employment',
    start-date: [... ]
  }
},
{
  {
    person: {...},
    company: {...}
    employment: { ... }
  }, ...
]
```

2) we also allow retrieving a matched attributes directly:

```
match 
$x isa person, has email $e; $e = "abc@gmail.com";
fetch
$e as person-email;
$x as person: name;
```

returns:

```
[
  {
    person-email: { 'type' : 'email', value: "abc@gmail.com" }
    person: { 
      type: 'person',
      name: [ {...}, {...}]
  },
  { 
    ...
  }
]
```

3) Structing nested fetch queries:

```
match 
$x isa person, has email $e; $e = "abc@gmail.com";
fetch
$x as person: name;
children: {
  match
  ($x, $c) isa parentship;
  $c isa person;
  fetch
  $c as child: name;
};
```

returns:

```
[
  {
    person: {
      type: 'person',
      name: [ { type: 'name', value: 'john'}, {type: 'nickname', value: 'johnny'}],
   },
   children: [
     {
        child: {
          type: 'person',
          name: [ {type: 'name', value: 'Alice'}]
        }
      },
      {
         child: {
          type: 'person',
          name: [ {type: 'name', value: 'Billy'}]
        }
      }, 
      ...
    ]
  },
  {
    ...
  }
]
```

In other words, `fetch` returns a set JSON schema that can be defined as:

```
FETCH_ANSWER_SCHEMA :=
  $attr as key1;  ==> creates entry key1: attribute-json
  $thing as key2: attr-1, attr2...; ==> creates entry key2: { type: <thing type>, attr-1: [ attribute-json, attribute-json...], attr-2: [attribute-json, attribute-json...], ...}
  key3: { match ... fetch k ... };  ==>  create entry key3: [ FETCH_ANSWER_SCHEMA  ] 
```

**Update**
The final agreed-upon response format has changed from the above proposal. Please see expected input/output pairs at [our Fetch BDD scenarios](https://github.com/vaticle/typedb-behaviour/blob/518e6ace0413c48f70903bded7ea5cb518d44170/query/language/fetch.feature)

**Comment by flyingsilverfin** [comment]
During this implementation we'll probably also implement https://github.com/vaticle/typeql/issues/284

**Comment by shreyas1599** [comment]
@flyingsilverfin had a question on this. Apologies if this is not the right place to post the question or if it's been answered elsewhere.

```
fetch
$x as person: name, age, address;  # fetch names, ages, and addresses for a the person
$y as company: address;            # fetch addresses for the company
$r as employment: start-date;      # fetch start-dates for the company
```

when this is run - if a person has only the `name` and `age` attributes defined but not the `address` will that particular instance be returned in the response with `address` as `undefined` ?

**Comment by flyingsilverfin** [comment]
This is an excellent question and actually spurred adding a new test case & improved error message to our type validation - thank you!

As an answer:
1) if your projection variable eg. `$x` _cannot_ own (according to the schema) the attribute type listed (or any of its subtypes), then the query will be rejected. This falls in line with our static type checking wherever possible philosophy, and is will generally direct user to write better more concise, accurate queries while avoiding bugs.
2) under the assumption that `$x` is allowed to own the attributes or some of the attributes' subtypes, you will receive an empty list in the return structure for those attributes if none are present:
```
[ 
{
  person: {
    type: 'person',   // guaranteed key
    name: [],
    age: [],
    address: []
  },
  company: {
    type: 'nonprofit' // respects polymorphism
    address: []
  },
  employment: {
    type: 'employment',
    start-date: []
  }
},
...
]
```
is a valid response from 'fetch'.

Note that the specification for return structure & queries has moved on a bit since this issue - I'll refer you to the the test suite here: https://github.com/vaticle/typedb-behaviour/blob/518e6ace0413c48f70903bded7ea5cb518d44170/query/language/fetch.feature

**Comment by shreyas1599** [comment]
Thanks @flyingsilverfin! Just to make sure I understood this right

> 2\. under the assumption that `$x` is allowed to own the attributes or some of the attributes' subtypes, you will receive an empty list in the return structure for those attributes if none are present:

This is a possibility then?

```
person: {
    type: 'person',   // guaranteed key
    name: [{ some entries here }],
    age: [],
    address: [{ some entries here}]
  },
```

**Comment by codeabuser007** [comment]
@flyingsilverfin Hi
Can you please provide at least one example how to use Fetch query with Node.js or any other languages.
You documentation lacking any explanation apart from how to use it within TypeDB instance

https://typedb.com/docs/typeql/data/fetch
https://typedb.com/docs/drivers/nodejs/api-reference#_QueryManager_fetch__query_string__options_TypeDBOptions

These are the docs I managed to find, but that doesnt show what method to use within a nodejs
However running .fetch() method within a typedb-driver returns some `FlatMapped` Class which no idea what to do with that value next

**Comment by codeabuser007** [comment]
@izmalk thanks a lot.
